server.py

```diff
@@ -36,32 +36,36 @@ def index():
 ######################################################################
 @app.route('/products', methods=['GET'])
 def list_products():
-    category        = request.args.get('category')
-    discontinued    = str2bool(request.args.get('discontinued'))
-    min_price       = request.args.get('min-price')
-    max_price       = request.args.get('max-price')
-    price           = request.args.get('price')
-    limit           = request.args.get('limit')
-    results         = []
-    count           = 0
-    cutoff          = 10
-
-    if (limit is not None and int(limit) >= 0):
-        cutoff = int(limit)
-    elif (limit is not None and int(limit) < 0):
-        return reply(results, HTTP_400_BAD_REQUEST)
-        
-    for key, value in products.iteritems():
-        if (count == cutoff):
-            break
-        if matches_clause(category,      value['category']) and \
+	category        = request.args.get('category')
+	discontinued    = str2bool(request.args.get('discontinued'))
+	min_price       = request.args.get('min-price')
+	max_price       = request.args.get('max-price')
+	price           = request.args.get('price')
+	limit           = request.args.get('limit')
+	search_term		= request.args.get('search')
+	results         = []
+	count           = 0
+	cutoff          = 10
+	
+	if (limit is not None and int(limit) >= 0):
+		cutoff = int(limit)
+	elif (limit is not None and int(limit) < 0):
+		return reply(results, HTTP_400_BAD_REQUEST)
+	for key, value in products.iteritems():
+		if (count == cutoff):
+			break
+		if matches_clause(category,      value['category']) and \
            matches_clause(discontinued,  value['discontinued']) and \
-           matches_price(min_price, max_price, price, value['price']):
-                results.append(products[key])
-        count += 1
-
-    return reply(results, HTTP_200_OK)
-
+           matches_price(min_price, max_price, price, value['price']) and\
+		   matches_term(search_term, value['category'], value['name'], value['description'].decode('utf-8')):
+			results.append(products[key])
+			count += 1
+	
+	return reply(results, HTTP_200_OK)
+
+def matches_term(term, category, name, description):
+	return (term is None or term.lower() in category.lower() or term.lower() in name.lower() or term.lower() in description.lower())
```

Comment:
When using the the `in` operator against a string, it will look for a substring. When using `in` for a collection/iterable, it will look for a matching element:
```
print(str('elec' in 'electric'))#True
print(str('elec' in ['electric']))#False
print(str('electronic' in ['electronic', 'appliances']))#True
```